### ISSUE TITLE
`jenkins` ruleset doesn't work when using Java 21

### ISSUE THREAD

**Issue by nre-ableton** [CodeNarc/CodeNarc#782]
When I try using the new `jenkins` ruleset introduced in CodeNarc 3.5.0 with Java 21, I get an error:

```
ERROR: org.codenarc.analyzer.AnalyzerException: Error analyzing source file: Jenkinsfile; BUG! exception in phase 'semantic analysis' in source unit 'None' Unsupported class file major version 65
org.codenarc.analyzer.AnalyzerException: Error analyzing source file: Jenkinsfile; BUG! exception in phase 'semantic analysis' in source unit 'None' Unsupported class file major version 65
```

The ruleset works fine when using Java 17.

**Comment by lslonina** [comment]
Faced same issue, it can be due to groovy and java 21 version support.
As checked Java 21 constant was added in version 4.0.11: https://issues.apache.org/jira/browse/GROOVY-10986

**Comment by chrismair** [comment]
> Codenarc has groovy 4.0.7. So small upgrade is required (here or in your build script).

Thanks for that info. Have either of you tried explicitly specifying the later Groovy version in your build/classpath to override the default 4.0.7? Does that fix it?

**Comment by lslonina** [comment]
Yes, I have used newest available groovy 4 dependency. All works fine. I'm on gradle and upgraded as specified in: [CodeNarc Gradle](https://docs.gradle.org/current/userguide/codenarc_plugin.html#sec:codenarc_dependency_management)

**Comment by chrismair** [comment]
Excellent. I'm glad there is a simple workaround. I'll plan to update the formal dependency in CodeNarc for the next release.

**Comment by nre-ableton** [comment]
@chrismair sounds good. Should I close this issue as "wontfix" or let you close it with a corresponding documentation "fix"?

**Comment by nre-ableton** [comment]
Hmm, actually, now that I think about it, I'm not sure that this is a great solution for this problem. While this workaround is great for developers, it won't work well on Jenkins itself, where the Groovy version is (sadly) stuck at 2.4.21.

Is there any way that we can avoid the JDK21 constant in the `jenkins` ruleset itself?

**Comment by chrismair** [comment]
I'll leave this open as documentation until the next release, and will upgrade the Groovy version then.

**Comment by nvuillam** [comment]
@chrismair what are the recommended versions of java & groovy to run CodeNarc ?
We have a similar issue with npm-groovy-lint :)
cc @stevenh

**Comment by nre-ableton** [comment]
> @chrismair what are the recommended versions of java & groovy to run CodeNarc ? We have a similar issue with npm-groovy-lint :) cc @stevenh

Theoretically, anything >=4.0.11 should work fine.

**Comment by nvuillam** [comment]
> > @chrismair what are the recommended versions of java & groovy to run CodeNarc ? We have a similar issue with npm-groovy-lint :) cc @stevenh
> 
> Theoretically, anything >=4.0.11 should work fine.

We ship with groovy 3.x ... thanks for the info :)

**Comment by chrismair** [comment]
I have reproduced the problem and confirmed that it appears to be fixed when using Groovy 4.0.15 and Groovy 3.0.22.

**Comment by chrismair** [comment]
Will be available in CodeNarc 3.6.0
